File: schedjuice-reimagined-be/app_scholis/teacher_sso.py
"""
Signing a teacher into Scholis from here.

A teacher who is already signed into Schedjuice should not have to prove it again
to Scholis. This mints the single-use link that does it; the browser follows the
link and Scholis exchanges it for its own session cookie.

Two things this deliberately does not do.

It cannot create an account. Scholis refuses a teacher it has never heard of, and
that refusal is correct: sign-in authenticates, it never provisions. A teacher
with no Scholis user has to be invited through Scholis's own path first, so the
error here says that rather than pretending the link is broken.

It does not keep an audit row of its own. Scholis already records who minted the
ticket (``minted_by_client_id``), the caller's own staff id (``externalRef``, sent
below), and when it was redeemed -- which is a better audit than a copy here,
because it includes the redemption this side never sees. Duplicating it would mean
two records that can disagree.
"""
from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import datetime

from app_scholis.errors import ScholisError, ScholisNotFoundError
from app_scholis.times import parse_when
from app_scholis.provisioning import org_client, require_active_connection

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class TeacherLink:
    url: str
    expires_at: datetime | None
    email: str


class TeacherNotAtScholis(ScholisError):
    """The address is fine; Scholis has no such teacher, so there is nobody to sign in."""


def mint_teacher_link(*, teacher=None, email: str | None = None) -> TeacherLink:
    """
    Mint a sign-in link for one teacher.

    ``teacher`` is preferred over a bare address: it supplies both the email and
    the ``externalRef`` that Scholis records against the ticket, which is what
    lets their side answer "who did Schedjuice think this was?"
    """
    address = _resolve_email(teacher, email)
    connection = require_active_connection()
    client = org_client(connection)

    external_ref = str(teacher.pk) if teacher is not None and teacher.pk else None

    try:
        result = client.mint_teacher_sso(email=address, external_ref=external_ref)
    except ScholisNotFoundError as e:
        # Scholis answers 404 both for "no such teacher" and for "a teacher at
        # another school", on purpose -- telling them apart would let a caller
        # enumerate which addresses hold accounts elsewhere. So this message must
        # not claim to know which it was.
        raise TeacherNotAtScholis(
            "Scholis has no teacher at that address. They need to be invited at "
            "Scholis before they can be signed in from here."
        ) from e

    url = result.get("url")
    if not isinstance(url, str) or not url:
        raise ScholisError("Scholis did not return a sign-in URL.")

    logger.info("Scholis: minted teacher sign-in link for %s", address)
    return TeacherLink(
        url=url, expires_at=parse_when(result.get("expiresAt")), email=address
    )


def _resolve_email(teacher, email: str | None) -> str:
    address = (getattr(teacher, "email", "") or email or "").strip()
    if not address:
        raise ScholisError("A teacher email is required to mint a sign-in link.")
    # Trimmed here rather than left to Scholis. Scholis lower-cases before
    # matching, so case does not matter, but its schema rejects surrounding
    # whitespace outright with a 400 -- which reads like a broken integration to
    # somebody whose address came out of a spreadsheet with a trailing space.
    return address


__all__ = ["TeacherLink", "TeacherNotAtScholis", "mint_teacher_link"]

File: apps/api/server/integrations/teacher-sso.ts
import { insertTeacherSsoTicket } from '@/data/teacher-sso';
import { findUserByEmail } from '@/data/users';
import type { AuthedContext } from '@/server/context.types';
import { forbidden, notFound } from '@/server/errors';
import { randomBytes } from 'node:crypto';
import { z } from 'zod';

/**
 * Mint a one-time link that signs an existing teacher in.
 *
 * The staff-side counterpart to `launchAttempt`. An integrating system already
 * knows who this teacher is — they are logged into it — so making them prove it
 * again with a magic link is a second inbox to check for no security gained.
 * The caller presents its org credential, Scholis hands back a link, the
 * teacher's browser follows it and arrives with a session.
 *
 * What this deliberately does not do is create anybody. Sign-in authenticates,
 * it never provisions; `lib/auth.ts` enforces that on the database write itself
 * rather than trusting a plugin option. A teacher with no account here is
 * refused and has to be invited through the normal path.
 */

/**
 * Long enough to survive a redirect and a slow page load, short enough that a
 * link leaked into a browser history or a proxy log is stale before it is
 * useful. Shorter than a launch ticket (15 minutes), because that one has to
 * survive a class of thirty devices arriving over several minutes and this one
 * is followed immediately by the person it was minted for.
 */
const TICKET_TTL_MS = 5 * 60 * 1000;

export const teacherSsoInput = z.object({
  // Normalised in the service, matching sign-up and team: the schema says what
  // shape is acceptable, the service decides what identity means.
  email: z.email(),
  /**
   * The caller's own staff identifier. Recorded for audit and never used to
   * authorise anything — authorisation is the resolved Scholis user.
   */
  externalRef: z.string().trim().max(120).optional(),
});

export type TeacherSsoInput = z.infer<typeof teacherSsoInput>;

export interface TeacherSsoTicket {
  url: string;
  expiresAt: string;
}

export const mintTeacherSsoTicket = async (
  ctx: AuthedContext,
  input: TeacherSsoInput,
): Promise<TeacherSsoTicket> => {
  if (!ctx.actor.scopes.includes('sso:write')) {
    throw forbidden('This key cannot sign teachers in.');
  }

  const email = input.email.trim().toLowerCase();
  const user = await findUserByEmail(ctx.db, email);

  // One refusal for "no such teacher" and "a teacher at another school" alike.
  // Email is globally unique here, so without this an org credential could
  // enumerate which addresses hold accounts at other schools by watching
  // whether the message changed.
  //
  // The optional chain is the whole rule, not a shorthand for a null check: a
  // missing user yields `undefined`, which matches no org id, so both cases
  // fall into the same branch by construction rather than by remembering to
  // write two conditions.
  if (user?.orgId !== ctx.actor.orgId) throw notFound('Teacher');

  const now = ctx.now();
  const token = randomBytes(32).toString('base64url');
  const expiresAt = new Date(now.getTime() + TICKET_TTL_MS);

  await insertTeacherSsoTicket(ctx.db, {
    orgId: ctx.actor.orgId,
    // Resolved here, so the ticket names an account rather than an address that
    // could later belong to somebody else.
    userId: user.id,
    token,
    externalRef: input.externalRef ?? null,
    // For a machine actor `userId` carries the client id — see
    // resolve-client-actor.ts. A person minting one of these is possible but
    // unusual, and leaves the column null.
    //
    // `kind === 'client'` is doing real work here, not just tidying the audit
    // trail: `minted_by_client_id` is a uuid foreign key onto api_clients, and
    // a client actor's userId is the only kind guaranteed to be one. The same
    // field on a person is a users.id, which would satisfy the type and then
    // fail the constraint — attributing the ticket to a credential that never
    // existed.
    mintedByClientId: ctx.actor.kind === 'client' ? ctx.actor.userId : null,
    expiresAt,
  });

  // Built here rather than in the route, so the link a caller is handed and the
  // link Scholis will honour cannot drift apart.
  const webOrigin = (process.env.WEB_ORIGIN ?? 'http://localhost:3000').replace(/\/$/, '');
  return {
    url: `${webOrigin}/sso?ticket=${encodeURIComponent(token)}`,
    expiresAt: expiresAt.toISOString(),
  };
};
